### ISSUE TITLE
Q: List of all possible permissions?

### ISSUE THREAD

**Issue by creadone** [CanCanCommunity/cancancan#615]
Hi, guys

In the my project need list of all possible permissions for frontend where UI elements will showing/hiding according to true/false values received from backend for each record. It should be something like this:

```ruby
{
  'ModelName1' => [
    :create,
    :read,
    :update,
    :destroy
  ], 
  'ModelName2' => [
    :download,
    :approve,
    :decline
  ]
}
```
Getting the list for the current user is pretty easy `Ability.new(current_users).send :rules` or `self.rules` inside instance but not for rules like this:

```
can :read Model if user.admin?
```

I did not find a solution for several hours of reading these issues, stackoverflow and docs. They always not showing if user not admin. I trying to rewrite rules in the simple format but it is also not help me. Eventually I came up with a variant that meets the requirements but this variant looks like dirty trick and possibly cause problems in the future:

**ability.rb**
```ruby
class Ability
  include CanCan::Ability
  include PermissionHelper  # add helper

  def initialize(user)
    @user = user # Make user accessible for helper
    can :create, Model if user.admin?  # Define permissions
    [...]
  end

  # Group rules by models
  def to_list
    store = {}
    rules.each do |rule|
      next if rule.subjects.first == :all
      model = rule.subjects.first
      action = rule.actions.first
      store.merge!(model.name => []) unless store.has_key?(model.name)
      store[model.name].push(action)
    end
    store.transform_values! &:uniq
  end
end
```

**permission_helper.rb**
```ruby
module PermissionHelper
  def all_permissions
    fake_user = @user.class.new
    self.class.new(fake_user)
  end
end
```

In short, if we create instance of Ability with User instance, we got the list of all rules regardless of the user's role. Then I iterate over this array and resolve each rule by specific record.

I think that if rule for resolve need to get info from DB my variant won't work. Is there an easier and more widely used way to get what I want?

**Comment by coorasse** [comment]
Please refer to [Stackoverflow for questions](http://stackoverflow.com/questions/ask?tags=cancancan). Thank you 🙇 .

**Comment by schmijos** [comment]
I think this is a useful reference for this issue: https://stackoverflow.com/questions/76574913/cancancan-get-all-dynamic-permissions
